sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Prefixing.hs

```diff
@@ -20,100 +21,102 @@ module DA.Cli.Damlc.Command.MultiIde.Prefixing (
+removeProgressTokenPrefix (LSP.ProgressTextToken (T.uncons -> Just ('t', rest))) =
+  bimap LSP.ProgressTextToken progressTokenPrefixFromMaybe $ progressTokenSplitPrefix rest
+removeProgressTokenPrefix t = (t, NoPrefix)
+
+-- Prefixes the SWindowWorkDoneProgressCreate and SProgress messages from subIDE. Rest are unchanged.
+addProgressTokenPrefixToServerMessage :: T.Text -> LSP.FromServerMessage -> LSP.FromServerMessage
+addProgressTokenPrefixToServerMessage prefix (LSP.FromServerMess LSP.SWindowWorkDoneProgressCreate req) =
+  let prefixedToken = addProgressTokenPrefix (SubIDEPrefix prefix) $ req ^. LSP.params . LSP.token
+   in LSP.FromServerMess LSP.SWindowWorkDoneProgressCreate $ req & LSP.params . LSP.token .~ prefixedToken
+addProgressTokenPrefixToServerMessage prefix (LSP.FromServerMess LSP.SProgress notif) =
+  case removeProgressTokenPrefix $ notif ^. LSP.params . LSP.token of
+    -- ProgressToken was created by this subIDE, add its usual prefix
+    (unprefixedToken, NoPrefix) ->
+      let prefixedToken = addProgressTokenPrefix (SubIDEPrefix prefix) unprefixedToken
+       in LSP.FromServerMess LSP.SProgress $ notif & LSP.params . LSP.token .~ prefixedToken
+    -- ProgressToken was created by client, send back the unprefixed token
+    (unprefixedToken, ClientPrefix) -> LSP.FromServerMess LSP.SProgress $ notif & LSP.params . LSP.token .~ unprefixedToken
+    (_, SubIDEPrefix t) -> error $ "SubIDE with prefix " <> T.unpack t <> " is somehow aware of its own prefixing. Something is very wrong."
+addProgressTokenPrefixToServerMessage _ msg = msg
+
+-- Prefixes client created progress tokens for all requests that can create them.
+addProgressTokenPrefixToClientMessage :: LSP.FromClientMessage' a -> LSP.FromClientMessage' a
+addProgressTokenPrefixToClientMessage = \case
   mess@(LSP.FromClientMess method params) ->
     case LSP.splitClientMethod method of
       LSP.IsClientReq -> do
         let progressLenses = getProgressLenses method
-        params' <- 
-          maybe pure (\lens -> runLens lens $ traverse addClientProgressTokenPrefix) (workDoneLens progressLenses) params
-            >>= maybe pure (\lens -> runLens lens $ traverse addClientProgressTokenPrefix) (partialResultLens progressLenses)
-        pure $ LSP.FromClientMess method params'
-      _ -> pure mess
-  rsp@LSP.FromClientRsp {} -> pure rsp
-  where
-    addClientProgressTokenPrefix :: LSP.ProgressToken -> IO LSP.ProgressToken
-    addClientProgressTokenPrefix unPrefixedToken = do
-      let prefixedToken = addProgressTokenPrefix "client" unPrefixedToken
-      -- We insert to the mapping backwards, as the mapping goes from server token to client token, and for client -> server messages, this should add, not remove, the prefix.
-      modifyMVar_ tokensVar (pure . Map.insert (prefixedToken, Nothing) unPrefixedToken)
-      pure prefixedToken
-
+            doAddProgressTokenPrefix = maybe id $ \lens -> runLens lens %~ fmap (addProgressTokenPrefix ClientPrefix)
```

Comment:
A type signature would be helpful here, I think.

You can compose `mapped` with `runLens lens` here, i.e. `runLens lens . mapped %~ ...` instead of `runLens lens %~ fmap ...` but that's just a stylistic choice.

Reply:
Allows me to drop some brackets, i like it